provisioners/bootstrap_posix.sh

```diff
@@ -0,0 +1,197 @@
+OPTIND=1
+
+# Parse options. Use of -l flags the script as non-interactive
+while getopts ":h?l:" opt; do
+    case "$opt" in
+        h|\?)
+            echo "Usage: ./bootstrap_posix.sh -h               - Show help"
+            echo "       ./bootstrap_posix.sh -l /path/logfile - Log output to file"
+            echo "       ./bootstrap_posix.sh                  - Interactive mode"
+            exit 0
+            ;;
+        l)
+            LOG_PATH=$OPTARG
+            NONINTERACTIVE=true
+            # touch log file to see if we have write access and fail otherwise
+            touch "$LOG_PATH" || fail "Can't write log to ${LOG_PATH}"
+            exec >"$LOG_PATH" 2>&1
+            ;;
+        :)
+            echo "Option -$OPTARG requires an argument" >&2
+            exit 1
+            ;;
+    esac
+done
+
+# Determine OSTYPE so we can set OS specific paths and alter logic if need be
+case "${OSTYPE}" in
+  darwin*)  OS='darwin' ;;
+  linux*)   OS='linux' ;;
+  *)        fail "OS either not detected or not supported!" ;;
+esac
+
+# Linux and Darwin share some common paths
+if [ $OS == "linux" ] || [ $OS == "darwin" ]; then
+    ROLE_FILE='/etc/puppet_role'
+    PUPPET_BIN='/opt/puppetlabs/bin/puppet'
+    FACTER_BIN='/opt/puppetlabs/bin/facter'
+    R10K_BIN='/opt/puppetlabs/bin/r10k'
+fi
+
+# This file should be set by the provisioner and is an error to not have a role
+# It indicates the role this node is to play
+# We may completely change the logic in determine a nodes role such as using an ENC
+# but for now, this works
+if [ -f "${ROLE_FILE}" ]; then
+    ROLE=$(<${ROLE_FILE})
+else
+    fail "Failed to find puppet role file ${ROLE_FILE}"
+fi
+
+# Check that we have the minimum requirements to run puppet
+# Since this is a bootstrap script we may actaully install minimum requirements here in the future
+if [ ! -x "${PUPPET_BIN}" ]; then
+    fail "${PUPPET_BIN} is missing or not executable"
+fi
+
+if [ ! -x "${FACTER_BIN}" ]; then
+    fail "${FACTER_BIN} is missing or not executable"
+fi
+
+if [ ! -x "${R10K_BIN}" ]; then
+    fail "${R10K_BIN} is missing or not executable"
+fi
+
+# Create a temp dir for executing puppet
+TMP_PUPPET_DIR=$(mktemp -d -t puppet_working)
+[ -d "${TMP_PUPPET_DIR}" ] || fail "Failed to mktemp puppet working dir"
+
+# Download puppet repository and extract
+function get_puppet_repo {
+    TMP_DL_DIR=$(mktemp -d -t puppet_download)
+    [ -d "${TMP_DL_DIR}" ] || fail "Failed to mktemp download dir"
+
+    # Download the puppet repo tarball directly from github
+    # We don't use git because some oses don't have git installed by default
+    # In the future, we may publish master branch to s3 or some other highly available service since
+    # Github has rate limits on downloads.
+    while true; do
+        echo "Downloading puppet repo: ${PUPPET_REPO}"
+        if HTTP_RES_CODE=$(curl -sL $PUPPET_REPO -o "${TMP_DL_DIR}/puppet.tar.gz" -w "%{http_code}") && [[ $HTTP_RES_CODE = "200" ]]; then
```

Comment:
Could you use curl's retry and back-off instead or in addition to this?

Reply:
I could but at this time I want the logic handled in the shell script itself.  By having it in the shell script, I don't worry about the possibility of differences in the versions of curl from system to system.  I can also retry endlessly and/or send out events from the script between retries.  And last, I intend to have different retry logic based upon whether the script is called interactively or not.

Reply:
sounds good. it is simpler to read like this too